### PR TITLE
[SPARK-32006][SQL] Create date/timestamp formatters once before collect in `hiveResultString()`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
1. Add method `getTimeFormatters` to `HiveResult` which creates timestamp and date formatters.
2. Move creation of `dateFormatter` and `timestampFormatter` from the constructor of the `HiveResult` object to `HiveResult. hiveResultString()` via `getTimeFormatters`. This allows to resolve time zone ID from Spark's session time zone `spark.sql.session.timeZone` and create date/timestamp formatters only once before collecting `java.sql.Timestamp`/`java.sql.Date` values.
3. Create date/timestamp formatters once in SparkExecuteStatementOperation.

### Why are the changes needed?
To fix perf regression comparing to Spark 2.4

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
- By existing test suite `HiveResultSuite` and etc.
- Re-generate benchmarks results of `DateTimeBenchmark` in the environment:

| Item | Description |
| ---- | ----|
| Region | us-west-2 (Oregon) |
| Instance | r3.xlarge |
| AMI | ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-20190722.1 (ami-06f2f779464715dc5) |
| Java | OpenJDK 64-Bit Server VM 1.8.0_252 and OpenJDK 64-Bit Server VM 11.0.7+10 |

### CODE REVIEW

**Comment by MaxGekk** [comment]
@cloud-fan @juliuszsompolski @HyukjinKwon @bogdanghit @gatorsmile Please, take a look at this PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala [hunk 1]
```diff
@@ -72,47 +90,32 @@ object HiveResult {
...
-    // Use `FastDateFormat` as the legacy formatter because it is thread-safe.
-    legacyFormat = LegacyDateFormats.FAST_DATE_FORMAT,
-    isParsing = false)
-  private def timestampFormatter = TimestampFormatter.getFractionFormatter(
-    DateTimeUtils.getZoneId(SQLConf.get.sessionLocalTimeZone))
-
   /** Formats a datum (based on the given data type) and returns the string representation. */
-  def toHiveString(a: (Any, DataType), nested: Boolean = false): String = a match {
+  def toHiveString(
+    a: (Any, DataType),
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/HiveResult.scala:95]: nit: 4 space indentation

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
The last commit updates indentation only, I'm merging it to master, thanks!

**Review by HyukjinKwon — COMMENTED** [review]
Ah, late LGTM

**Pull request closed without merging**